csrc/multidevice/utils.cpp

```diff
@@ -177,6 +177,21 @@ bool isResharding(const Expr* expr) {
     return false;
   }
 
+  // Reduction over a sharded dimension.
```

Comment:
The sharding propagation pass will propagate DIDx through an AllReduce
DIDx(i0), i1 -> r(DIDx(i0), i1. Since `isResharding` is used to detect when an expr is lowered to a collective, the AllReduce would not get identified without explicitly checking for a reduction of a DID axis.

Reply:
cc @naoyam 

This is related to our broadcast/DID discussion last week. I've yet to understand the difference between `r(DIDx{i0})` and `r(i0)`. 
1. Are they semantically equivalent? 
2. If so, should nvFuser support both or only a canonical form (can be either of them)?

Reply:
First of all, one caveat is that I remember there's a slight semantic difference of interpreting parallel types between DID and BID/TID. I don't exactly remember what it is, though. Let me check, but @samnordmann, I remember I told you. Do you remember?

So, suppose we have `r(TIDx{i0})` and `r(i0)`. They are different that:

- Only the former instructs nvFuser to parallelize the reduction. In this case, the `blockReduce` device function defined in one of the runtime files will be used.
- The latter indicates serial execution. Semantically, it should be valid if only one of the threads, e.g., tid==0, runs the reduction. I think by default we just let all threads run the reduction redundantly. There's a predicate generation pass to limit the redundant execution, so that may affect.
- In the latter case, since it's executed by just a single thread, all the elements must be accessible by the thread. If the input is parallelized by TIDx, it means the input must be on shared memory. A syncthreads will be automatically inserted.

BID is also the same.

Reply:
We've had many conversation around this.

`DIDx(i0) -> ri0` vs. `DIDx(i0) -> r(DIDx(i0))`
Since we don't codegen the collectives, we can't control how the parallelization occurs, so the two would both get lowered to an (all)reduce collective. In a sense the DID parallelization is only conveying the state of the axis.


ReduceScatter is not represented well with `r(DIDx(i0))` unfortunately... 
Currently we look for a pattern like this: `[DIDx(i0), i1] -> [r(i0), DIDx(i1)]`

If we use `r(DIDx(i0))` then we get: `[DIDx(i0), i1] -> [r(DIDx(i0)), DIDx(i1)]`
We have 2 DIDx axis, which is not allowed, but in this case the `r(DIDx(i0))` is only conveying the reduction happened on a sharded axis.


There was another discussion with @naoyam where we discussed that ideally we should represent reduce scatter as two primitives (reduce + scatter), but since it would complicate the communication lowering pass we did not do this.

Reply:
Thank you both! 

> Since we don't codegen the collectives, we can't control how the parallelization occurs, so the two would both get lowered to an (all)reduce collective

That's a great summary. 

Back to my earlier question: 
1. `DIDx(i0) -> ri0` and `DIDx(i0) -> r(DIDx(i0))` are indeed semantically equivalent, although not the case for TID/BID which affects codegen. 
2. For the time being, we need the former to avoid two `DIDx`s for ReduceScatter. Also, [newForReduction](https://github.com/NVIDIA/Fuser/blob/2edebb60130b6f17c5b328279b41d35c15a21fbd/csrc/ops/arith.cpp#L1197-L1203), which I changed weeks ago, doesn't propagate DIDs to reduction dimensions and therefore implements the former. (It's easy to change that either way.) So should we stick with only the former (for other ops) or should we support both?

Reply:
Hmm.. I don't have strong opinions, but maybe we should stick to the former to stay consistent?

Reply:
> DIDx(i0) -> ri0 vs. DIDx(i0) -> r(DIDx(i0))
> Since we don't codegen the collectives, we can't control how the parallelization occurs, so the two would both 
> get lowered to an (all)reduce collective. In a sense the DID parallelization is only conveying the state of the axis.

> DIDx(i0) -> ri0 and DIDx(i0) -> r(DIDx(i0)) are indeed semantically equivalent

I don't follow what you meant. In the former case, the output is not parallelized, so it indicates the operation does not need to be parallelized. How they are interpreted has some flexibility within the semantics, so they could result in the same execution, but that doesn't mean there no semantic difference. 

If the output is not parallelized, all of the below cases should be valid:

1. All of the GPUs execute `allreduce` collectively. All GPUs will have the valid reduction result. This is what we do now with DID.
2. All of the GPUs execute `reduce` collectively. Only one of the GPUs will have the valid result. Can be much more efficient than case 1.
3. It is only executed by one of the GPUs. The rest do nothing. This is not possible with distributed tensors.

In TID/BID, when the output is not parallelized, it is interpreted as case 3. If the input is TID parallelized, it must be placed on the shared memory or global memory. When it's parallelized, it's interpreted as case 2.

The case 1 behavior is actually `reduction + broadcast`. The reduction removes the reduced iter domain, so it no longer indicates nothing about the sharding, so a reduction itself does not indicate the behavior like case 1 but needs a broadcast. For example,

```
// t0: [DID(i0)]
t1 = sum(t0, {0})
t1->axis(0)->parallelize(DID)
// t1: [DID(r0)]
t2 = set(t1)
// t2: [] 
```

`t2` is a scalar tensor. It doesn't need to be sharded over DID.  The above fusion only instructs a reduction to a device. That means `MPI_Reduce` is enough.

`MPI_Allreduce` corresponds to below:

```
// t0: [DID(i0)]
t1 = sum(t0, {0})
t1->axis(0)->parallelize(DID)
// t1: [DID(r0)]
t2 = broadcast(t1, {true})
t2->axis(0)->parallelize(DID)
// t2: [DID(b0)] 
```

This pattern appears in normalization kernels, for which we have a lowering [pass](https://github.com/NVIDIA/Fuser/blob/main/csrc/device_lower/analysis/fused_reduction.cpp) that converts `reduce + broadcast` to `allreduce`. 

To be clear, I'm not saying the above behavior is absolutely right, but it's the behavior of the TID/BID parallelization and all of the existing analyses enabling the two parallel types have been built upon the behavior. For prototyping, I don't think we should care much, however, what concerns me most is, when integrating DID more deeply, these semantic differences might bite us in an unexpected, hard-to-understand way.

Reply:
> case 2: All of the GPUs execute reduce collectively. Only one of the GPUs will have the valid result. Can be much more efficient than case 1.

We also need to remember the device mesh is involved which I left out since we are mostly dealing with same device mesh. So in this example if the device mesh between the input and output is the same, then we all the GPUs must have the final result which would make case 2. If the output mesh is a single device (e.g. something like mesh moves from {0,1,2} -> {0} then we can use a single Reduce.

Reply:
>  then we all the GPUs must have the final result 

Why?

Reply:
> > then we all the GPUs must have the final result

> Why?

Because otherwise nvFuser wouldn't know which GPU holds the result. 

Below is my understanding of the semantics of a **TensorView**, which agrees with @cowanmeg . 
1. `r(DID{i0}), mesh={0}` means the reduction result is only valid on GPU 0. 
2. `r(i0), mesh={0}` means the same as 1. 
3. `r(DID{i0}), mesh={0,1,2,3}` means the reduction result is replicated on all four GPUs. Assume `i0=4` for now. 
4. `r(i0), mesh={0,1,2,3}` means the same as 3. 

That being said, I think you are raising a good question of whether an IterDomain is parallelized on DID may affect the semantics of the **op**. For example, we can use `DID{i0} -> r(i0)` to represent an `Allreduce`, and `DID{i0} -> r(DID{i0})` a `Reduce+Broadcast`. If so, nvFuser will have to support both. However, there's also the approach to represent `Reduce+Broadcast` as 
```
DID{i0}, mesh={0,1,2,3} -> [reduction] r(i0), mesh={0}
r(i0), mesh={0} -> [set] -> (the IterDomain no longer exists in the output), mesh={0,1,2,3}
```
which I believe nvFuser already supports today.

Reply:
> Because otherwise nvFuser wouldn't know which GPU holds the result.

Hmm, in the case of TID/BID, it's just tid==0 or bid==0. In the case of DID, since we specify the head device, doesn't we know where it is?

Reply:
> head device

I haven't heard about that notion, but we certainly could assume the first device in the mesh is the head device. 

Let's talk about this tomorrow? I believe the semantics you proposed can work. If we go with that, code that currently generates `b(DIDx{i0})` or `r(DIDx{i0})` will need to be changed to generate `b(i0)` or `r(i0)` instead.

Reply:
(In case it wasn't clear, this discussion has no intention to block the PR.)

Reply:
Happy to discuss tomorrow. For the record, the head device refers to the `root` in `ncclReduce`, for example. https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/api/colls.html#ncclreduce

Reply:
Even if we arbitrarily made the first device the root...

t0[DIDx(i0), i1] -> t1[r(i0), i1] -> t2[i1] -> t3[i1] all with mesh - {0,1,2}
t1 would only be reduced on gpu 0. How would t3 know if it is valid on the entire mesh or just gpu 0 without traversing back up to the root tensor looking for all reductions and seeing if it was an allreduce or reduce?

Reply:
also back to the PR...just to make sure we are on the same page for short term:
DIDx(i0) mesh -> r(i0) mesh --> same mesh lowered to AllReduce. 
DIDx(i0) mesh0 -> r(i0) mesh1 --> different mesh lowered to Reduce(s) where mesh1 makes up the roots

This means sharding propagation will not propagate DIDx through a reduction.

Reply:
> also back to the PR...just to make sure we are on the same page for short term:
DIDx(i0) mesh -> r(i0) mesh --> same mesh lowered to AllReduce.
DIDx(i0) mesh0 -> r(i0) mesh1 --> different mesh lowered to Reduce(s) where mesh1 makes up the roots

Yes. I believe that also matches our current implementation of communication lowering. 

> This means sharding propagation will not propagate DIDx through a reduction.

That seems safer -- it won't run into the arguable case of `r(DID{i0})`. 

As an incidental note, https://github.com/NVIDIA/Fuser/blob/13fba3b62c7b2230405c4b240b4bdaca6a363a1c/tests/cpp/test_multidevice_matmul.cpp#L87 generates `b(DID{i0})`, which is as arguable.

Reply:
haha yes, I actually remember specifically doing that to prevent an resharding expression from getting added (in this case it would insert an unecessary allgather)